nixos/modules/services/development/jupyterhub/default.nix

```diff
@@ -0,0 +1,117 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.jupyterhub;
+
+  jupyterhub_env = pkgs.python3.withPackages (p: with p; cfg.hubPackages);
+  jupyterlab_env = pkgs.python3.withPackages (p: with p; cfg.jupyterPackages);
+
+  jupyterhubConfig = pkgs.writeText "jupyterhub_config.py" ''
+    c.JupyterHub.bind_url = "http://${cfg.host}:${toString cfg.port}"
+
+    c.JupyterHub.authentication_class = "${cfg.authentication}"
+    c.JupyterHub.spawner_class = "${cfg.spawner}"
+
+    c.SystemdSpawner.cmd = "${jupyterlab_env}/bin/jupyterhub-singleuser"
+
+    ${cfg.extraConfig}
+  '';
+in {
+  meta.maintainers = with maintainers; [ costrouc ];
+
+  options.services.jupyterhub = {
+    enable = mkEnableOption "Jupyterhub development server";
+
+    authentication = mkOption {
+      type = types.str;
+      default = "jupyterhub.auth.PAMAuthenticator";
+      description = ''
+        Jupyterhub authentication to use
+      '';
+    };
+
+    spawner = mkOption {
+      type = types.str;
+      default = "systemdspawner.SystemdSpawner";
+      description = ''
+        Jupyterhub spawner to use
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.str;
+      default = "";
+      description = ''
+        Extra contents appended to the jupyterhub configuration
+      '';
+    };
```

Comment:
do you mind making a small example for this one, they help me immensely

Reply:
Created an example of setting memory and cpu limits on each users jupyterlab instance.